internal/controller/negotiation.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2020 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/redskyops/redskyops-controller/api/v1alpha1"
+	"github.com/redskyops/redskyops-controller/api/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/rest"
+)
+
+// MEGA HACK ALERT!!!
+
+// This code exists to force client side conversion between versions to avoid relying on webhook based server side
+// conversions (alpha in Kube 1.13). Instead of properly conversion the resources in storage, we look at the
+// representation provided by the API server and try to detect if we should override the `apiVersion`, thereby
+// initiating conversion.
+
+// WithConversion returns the supplied rest.Config with the negotiated serializer set so version conversion between
+// objects occurs during client-go reads into the cache. This ensures the controller never actually sees old
+// representations of the objects (and we lazily migrate storage to the latest representation).
+func WithConversion(config *rest.Config, scheme *runtime.Scheme) *rest.Config {
+	config.NegotiatedSerializer = &ConversionSerializer{
+		NegotiatedSerializer: serializer.WithoutConversionCodecFactory{
+			CodecFactory: serializer.NewCodecFactory(scheme),
+		},
+		scheme: scheme,
+	}
+	return config
+}
+
+// ConversionSerializer is a negotiated serializer that also handles the representation migration hack.
+type ConversionSerializer struct {
+	runtime.NegotiatedSerializer
+	scheme *runtime.Scheme
+}
+
+func (c *ConversionSerializer) DecoderToVersion(serializer runtime.Decoder, gv runtime.GroupVersioner) runtime.Decoder {
+	return &ConversionDecoder{Decoder: c.NegotiatedSerializer.DecoderToVersion(serializer, gv), scheme: c.scheme}
+}
+
+// ConversionDecoder is a decoder that decodes with the representation migration hack.
+type ConversionDecoder struct {
+	runtime.Decoder
+	scheme *runtime.Scheme
+}
+
+func (c *ConversionDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
```

Comment:
This method is invoked on each raw response coming back from the Kube API server. Typically, because of the way the controller-runtime cache works, this will we a `List` object (i.e. `ExperimentList`). For our objects this will check expected non-zero values before deserializing a second time (using the old version), thus allowing us to confirm that an object requires conversion to a newer representation. By injecting this logic here we ensure the controller never actually sees the old representations.